File: scripts/poll.py
"""Poll @<X_USERNAME>'s X profile via Scrapling and forward GitHub-bearing reposts to Gmail."""
from __future__ import annotations

import datetime as dt
import json
import os
import re
import smtplib
import sys
from email.mime.text import MIMEText

from scrapling.fetchers import StealthyFetcher

X_USERNAME = os.environ["X_USERNAME"]
X_AUTH_TOKEN = os.environ["X_AUTH_TOKEN"]
X_CT0 = os.environ["X_CT0"]
GMAIL_USER = os.environ["GMAIL_USER"]
GMAIL_APP_PASSWORD = os.environ["GMAIL_APP_PASSWORD"]
FORWARD_TO = os.environ.get("FORWARD_TO") or GMAIL_USER
STATE_FILE = os.environ.get("STATE_FILE", "state.json")
DRY_RUN = os.environ.get("DRY_RUN") == "1"
DEBUG = os.environ.get("DEBUG") == "1"

COOKIES = [
    {"name": "auth_token", "value": X_AUTH_TOKEN, "domain": ".x.com", "path": "/",
     "httpOnly": True, "secure": True, "sameSite": "None"},
    {"name": "ct0", "value": X_CT0, "domain": ".x.com", "path": "/",
     "httpOnly": False, "secure": True, "sameSite": "Lax"},
]

GH_RE = re.compile(r"https?://github\.com/[\w.-]+/[\w.-]+", re.IGNORECASE)
REPOST_RE = re.compile(r"repost|retweet|リポスト|リツイート", re.IGNORECASE)
STATUS_RE = re.compile(r"/status/(\d+)")


def load_state() -> dict:
    try:
        with open(STATE_FILE, encoding="utf-8") as f:
            return json.load(f)
    except (FileNotFoundError, json.JSONDecodeError):
        return {"processed_ids": []}


def save_state(state: dict) -> None:
    state["processed_ids"] = state.get("processed_ids", [])[-300:]
    state["last_run_at"] = dt.datetime.now(dt.UTC).isoformat()
    with open(STATE_FILE, "w", encoding="utf-8") as f:
        json.dump(state, f, indent=2, ensure_ascii=False)
        f.write("\n")


def fetch_tweets() -> list[dict]:
    print(f"Visiting https://x.com/{X_USERNAME}")
    page = StealthyFetcher.fetch(
        f"https://x.com/{X_USERNAME}",
        headless=True,
        network_idle=True,
        wait_selector='article[data-testid="tweet"]',
        wait_selector_state="visible",
        timeout=60000,
        cookies=COOKIES,
        locale="ja-JP",
        google_search=False,
        disable_resources=True,  # block fonts/images for speed; tweet text is text content
    )

    if page.status >= 400:
        raise SystemExit(f"X returned status {page.status}")

    articles = page.css('article[data-testid="tweet"]')
    out: list[dict] = []
    for a in articles:
        # tweet ID — first /status/<id> href inside the article
        status_links = a.css('a[href*="/status/"]::attr(href)') or []
        status_links = list(status_links)
        tweet_id = None
        for h in status_links:
            m = STATUS_RE.search(str(h))
            if m:
                tweet_id = m.group(1)
                break
        if not tweet_id:
            continue

        ctx_text = (a.css('[data-testid="socialContext"]::text').get() or "").strip()
        is_repost = bool(REPOST_RE.search(ctx_text))

        tweet_text_parts = a.css('[data-testid="tweetText"] *::text').getall() or []
        tweet_text = "".join(tweet_text_parts).strip()
        if not tweet_text:
            tweet_text = (a.css('[data-testid="tweetText"]::text').get() or "").strip()

        anchors = a.css("a")
        haystack_pieces: list[str] = []
        for el in anchors:
            for sel in ("::attr(href)", "::attr(aria-label)", "::text"):
                v = el.css(sel).get()
                if v:
                    haystack_pieces.append(str(v))
        haystack = " ".join(haystack_pieces)
        gh_matches = GH_RE.findall(haystack)

        out.append({
            "id": tweet_id,
            "isRepost": is_repost,
            "ctxText": ctx_text,
            "tweetText": tweet_text,
            "githubMatches": gh_matches,
        })
    return out


def normalize_gh_url(u: str) -> str | None:
    m = re.search(r"github\.com/([\w.-]+)/([\w.-]+)", u, re.IGNORECASE)
    if not m:
        return None
    owner, repo = m.group(1), m.group(2)
    repo = re.sub(r"\.git$", "", repo, flags=re.IGNORECASE)
    repo = re.sub(r"[.,;:]+$", "", repo)
    if owner.lower() in {"i", "search", "settings", "explore"}:
        return None  # X internal paths
    return f"https://github.com/{owner}/{repo}"


def send_one(tweet: dict, repos: list[str]) -> None:
    slug = repos[0].replace("https://github.com/", "")
    extra = f" (+{len(repos) - 1})" if len(repos) > 1 else ""
    subject = f"[repo-feed] {slug}{extra}"
    body = "\n".join([
        f"Reposted by @{X_USERNAME}",
        f"Tweet: https://x.com/i/status/{tweet['id']}",
        "",
        "Repos:",
        *[f"  {r}" for r in repos],
        "",
        "--- tweet text ---",
        tweet["tweetText"] or "(no text)",
    ])

    if DRY_RUN:
        print(f"[DRY] would send: {subject}")
        return

    msg = MIMEText(body, "plain", "utf-8")
    msg["Subject"] = subject
    msg["From"] = GMAIL_USER
    msg["To"] = FORWARD_TO

    with smtplib.SMTP_SSL("smtp.gmail.com", 465, timeout=30) as smtp:
        smtp.login(GMAIL_USER, GMAIL_APP_PASSWORD)
        smtp.send_message(msg)
    print(f"Sent: {subject}")


def main() -> None:
    state = load_state()
    seen = set(state.get("processed_ids", []))
    tweets = fetch_tweets()
    print(f"Fetched {len(tweets)} tweets from profile")
    repost_count = sum(1 for t in tweets if t["isRepost"])
    gh_count = sum(1 for t in tweets if t["githubMatches"])
    print(f"  reposts detected: {repost_count}, with github URL: {gh_count}")
    if DEBUG:
        for t in tweets:
            print(f'  tweet {t["id"]} repost={t["isRepost"]} ctx="{t["ctxText"]}" gh={len(t["githubMatches"])}')

    candidates = [t for t in tweets if t["isRepost"] and t["githubMatches"] and t["id"] not in seen]
    print(f"New GitHub-containing reposts: {len(candidates)}")

    for t in candidates:
        seen_repos: dict[str, None] = {}
        for u in t["githubMatches"]:
            n = normalize_gh_url(u)
            if n and n not in seen_repos:
                seen_repos[n] = None
        repos = list(seen_repos)
        if not repos:
            continue
        try:
            send_one(t, repos)
            state["processed_ids"].append(t["id"])
        except Exception as e:
            print(f"Failed {t['id']}: {e}", file=sys.stderr)

    save_state(state)
    print("Done.")


if __name__ == "__main__":
    main()

File: scripts/poll.mjs
import { chromium } from 'playwright';
import nodemailer from 'nodemailer';
import { readFile, writeFile } from 'node:fs/promises';
import { existsSync } from 'node:fs';

const X_USERNAME = process.env.X_USERNAME;
const X_AUTH_TOKEN = process.env.X_AUTH_TOKEN;
const X_CT0 = process.env.X_CT0;
const GMAIL_USER = process.env.GMAIL_USER;
const GMAIL_APP_PASSWORD = process.env.GMAIL_APP_PASSWORD;
const FORWARD_TO = process.env.FORWARD_TO || GMAIL_USER;
const STATE_FILE = process.env.STATE_FILE || 'state.json';
const DRY_RUN = process.env.DRY_RUN === '1';

for (const [k, v] of Object.entries({ X_USERNAME, X_AUTH_TOKEN, X_CT0, GMAIL_USER, GMAIL_APP_PASSWORD })) {
  if (!v) {
    console.error(`Missing required env: ${k}`);
    process.exit(1);
  }
}

async function loadState() {
  if (!existsSync(STATE_FILE)) return { processed_ids: [] };
  try { return JSON.parse(await readFile(STATE_FILE, 'utf-8')); }
  catch { return { processed_ids: [] }; }
}

async function saveState(state) {
  state.processed_ids = state.processed_ids.slice(-300);
  state.last_run_at = new Date().toISOString();
  await writeFile(STATE_FILE, JSON.stringify(state, null, 2) + '\n');
}

async function fetchTimeline() {
  const browser = await chromium.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-blink-features=AutomationControlled'],
  });
  const context = await browser.newContext({
    userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36',
    viewport: { width: 1280, height: 1800 },
    locale: 'ja-JP',
  });

  await context.addCookies([
    { name: 'auth_token', value: X_AUTH_TOKEN, domain: '.x.com', path: '/', httpOnly: true, secure: true, sameSite: 'None' },
    { name: 'ct0', value: X_CT0, domain: '.x.com', path: '/', httpOnly: false, secure: true, sameSite: 'Lax' },
  ]);

  const page = await context.newPage();
  console.log(`Visiting https://x.com/${X_USERNAME}`);
  await page.goto(`https://x.com/${X_USERNAME}`, { waitUntil: 'domcontentloaded', timeout: 45000 });

  try {
    await page.waitForSelector('article[data-testid="tweet"]', { timeout: 20000 });
  } catch {
    const html = await page.content();
    if (html.includes('Log in to X') || html.includes('ログイン')) {
      throw new Error('LOGIN_REQUIRED: cookies appear invalid or expired');
    }
    throw new Error('No tweets rendered within 20s');
  }

  for (let i = 0; i < 3; i++) {
    await page.evaluate(() => window.scrollBy(0, 1500));
    await page.waitForTimeout(1200);
  }

  const tweets = await page.$$eval('article[data-testid="tweet"]', (articles, viewer) => {
    return articles.map(a => {
      const statusLink = Array.from(a.querySelectorAll('a[href*="/status/"]'))
        .map(l => l.getAttribute('href'))
        .find(h => /\/status\/\d+/.test(h));
      const idMatch = statusLink?.match(/\/status\/(\d+)/);
      const id = idMatch?.[1];

      const ctxEl = a.querySelector('[data-testid="socialContext"]');
      const ctxText = (ctxEl?.textContent || '').trim();
      // On the user's own profile, socialContext indicates a repost.
      // Don't require the username inside ctxText — X sometimes drops it.
      const isRepost = /repost|retweet|リポスト|リツイート/i.test(ctxText);

      const tweetText = (a.querySelector('[data-testid="tweetText"]')?.textContent || '').trim();

      const hrefs = Array.from(a.querySelectorAll('a[href]')).map(l => l.getAttribute('href')).filter(Boolean);
      const expandedHrefs = Array.from(a.querySelectorAll('a[href]'))
        .map(l => l.getAttribute('aria-label') || l.textContent || l.getAttribute('href'))
        .filter(Boolean);
      const allHrefs = [...hrefs, ...expandedHrefs].join(' ');
      const githubMatches = allHrefs.match(/https?:\/\/github\.com\/[\w.-]+\/[\w.-]+/gi) || [];

      return { id, isRepost, ctxText, tweetText, githubMatches };
    });
  }, X_USERNAME);

  await browser.close();
  return tweets.filter(t => t.id);
}

function normalizeGithubUrl(u) {
  const m = u.match(/github\.com\/([\w.-]+)\/([\w.-]+)/i);
  if (!m) return null;
  const [, owner, repo] = m;
  return `https://github.com/${owner}/${repo.replace(/\.git$/i, '').replace(/[.,;:]+$/, '')}`;
}

async function sendOne(tweet, repos) {
  const transporter = nodemailer.createTransport({
    service: 'gmail',
    auth: { user: GMAIL_USER, pass: GMAIL_APP_PASSWORD },
  });

  const slug = repos[0].replace('https://github.com/', '');
  const subject = `[repo-feed] ${slug}${repos.length > 1 ? ` (+${repos.length - 1})` : ''}`;
  const body = [
    `Reposted by @${X_USERNAME}`,
    `Tweet: https://x.com/i/status/${tweet.id}`,
    '',
    'Repos:',
    ...repos.map(r => `  ${r}`),
    '',
    '--- tweet text ---',
    tweet.tweetText || '(no text)',
  ].join('\n');

  if (DRY_RUN) {
    console.log(`[DRY] would send: ${subject}\n${body}\n`);
    return;
  }
  await transporter.sendMail({ from: GMAIL_USER, to: FORWARD_TO, subject, text: body });
  console.log(`Sent: ${subject}`);
}

async function main() {
  const state = await loadState();
  const seen = new Set(state.processed_ids);
  const tweets = await fetchTimeline();
  console.log(`Fetched ${tweets.length} tweets from profile`);
  const repostCount = tweets.filter(t => t.isRepost).length;
  const ghCount = tweets.filter(t => t.githubMatches.length > 0).length;
  console.log(`  reposts detected: ${repostCount}, with github URL: ${ghCount}`);
  if (process.env.DEBUG === '1') {
    for (const t of tweets) {
      console.log(`  tweet ${t.id} repost=${t.isRepost} ctx="${t.ctxText}" gh=${t.githubMatches.length}`);
    }
  }

  const candidates = tweets
    .filter(t => t.isRepost && t.githubMatches.length > 0)
    .filter(t => !seen.has(t.id));

  console.log(`New GitHub-containing reposts: ${candidates.length}`);

  for (const t of candidates) {
    const repos = [...new Set(t.githubMatches.map(normalizeGithubUrl).filter(Boolean))];
    if (repos.length === 0) continue;
    try {
      await sendOne(t, repos);
      state.processed_ids.push(t.id);
    } catch (e) {
      console.error(`Failed ${t.id}: ${e.message}`);
    }
  }

  await saveState(state);
  console.log('Done.');
}

main().catch(e => {
  console.error('FATAL:', e);
  process.exit(1);
});
